File: mobile/ios/App/ShareStaging/ShareStagingStore.swift
import Foundation

/**
 * App Group durable staging for iOS Share → containing app upload.
 * Layout mirrors Android ShareStagingStore:
 *   <group>/intake-staging/<clientBatchId>/{meta.json, 000.jpg, …}
 *   <group>/intake-staging/pending.json
 */
enum ShareStagingStore {
  static let appGroupId = "group.co.rematcher.exchange"

  static var containerURL: URL? {
    FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: appGroupId)
  }

  static var stagingRoot: URL? {
    guard let container = containerURL else { return nil }
    let root = container.appendingPathComponent("intake-staging", isDirectory: true)
    try? FileManager.default.createDirectory(at: root, withIntermediateDirectories: true)
    return root
  }

  static func batchDirectory(clientBatchId: String) -> URL? {
    guard let root = stagingRoot else { return nil }
    return root.appendingPathComponent(clientBatchId, isDirectory: true)
  }

  static func loadPending() -> [String: String]? {
    guard let root = stagingRoot else { return nil }
    let url = root.appendingPathComponent("pending.json")
    guard let data = try? Data(contentsOf: url),
          let obj = try? JSONSerialization.jsonObject(with: data) as? [String: String]
    else { return nil }
    return obj
  }

  static func clearPending() {
    guard let root = stagingRoot else { return }
    try? FileManager.default.removeItem(at: root.appendingPathComponent("pending.json"))
  }

  static func loadMeta(clientBatchId: String) -> [String: Any]? {
    guard let dir = batchDirectory(clientBatchId: clientBatchId) else { return nil }
    let url = dir.appendingPathComponent("meta.json")
    guard let data = try? Data(contentsOf: url),
          let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    else { return nil }
    return obj
  }

  static func listImageFiles(clientBatchId: String) -> [URL] {
    guard let dir = batchDirectory(clientBatchId: clientBatchId),
          let files = try? FileManager.default.contentsOfDirectory(
            at: dir,
            includingPropertiesForKeys: [.fileSizeKey],
            options: [.skipsHiddenFiles]
          )
    else { return [] }
    let allowed: Set<String> = ["jpg", "jpeg", "png", "webp", "heic", "heif"]
    return files
      .filter { allowed.contains($0.pathExtension.lowercased()) }
      .sorted { $0.lastPathComponent < $1.lastPathComponent }
  }

  static func deleteBatch(clientBatchId: String) {
    if let dir = batchDirectory(clientBatchId: clientBatchId) {
      try? FileManager.default.removeItem(at: dir)
    }
    if let pending = loadPending(), pending["clientBatchId"] == clientBatchId {
      clearPending()
    }
  }
}

File: android/app/src/main/java/co/rematcher/exchange/share/ShareStagingStore.java
package co.rematcher.exchange.share;

import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.provider.OpenableColumns;
import android.util.Base64;
import android.util.Log;
import org.json.JSONArray;
import org.json.JSONObject;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

/** App-private durable staging for Share payloads before authenticated upload ACK. */
public class ShareStagingStore {
  private static final String TAG = "ShareStaging";
  private final Context context;

  public ShareStagingStore(Context context) {
    this.context = context.getApplicationContext();
  }

  private File root() {
    File d = new File(context.getFilesDir(), "intake-staging");
    if (!d.exists()) d.mkdirs();
    return d;
  }

  public File batchDir(String batchId) {
    File d = new File(root(), batchId);
    if (!d.exists()) d.mkdirs();
    return d;
  }

  /** Copy content URIs into durable app-private files and write meta.json. */
  public void save(String batchId, List<Uri> uris, String text) {
    File dir = batchDir(batchId);
    JSONArray files = new JSONArray();
    int index = 0;
    for (Uri uri : uris) {
      try {
        String mime = context.getContentResolver().getType(uri);
        if (mime == null) mime = "image/jpeg";
        String ext = mime.contains("png") ? "png" : mime.contains("webp") ? "webp" : "jpg";
        String displayName = queryDisplayName(uri);
        String name = String.format("%03d.%s", index, ext);
        File dest = new File(dir, name);
        try (InputStream in = context.getContentResolver().openInputStream(uri);
             FileOutputStream out = new FileOutputStream(dest)) {
          if (in == null) continue;
          byte[] buf = new byte[8192];
          int n;
          while ((n = in.read(buf)) >= 0) out.write(buf, 0, n);
        }
        JSONObject f = new JSONObject();
        f.put("name", name);
        f.put("originalName", displayName != null ? displayName : name);
        f.put("mime", mime);
        f.put("bytes", dest.length());
        files.put(f);
        index++;
      } catch (Exception e) {
        Log.w(TAG, "Failed to stage uri", e);
      }
    }

    try {
      JSONObject meta = new JSONObject();
      meta.put("clientBatchId", batchId);
      meta.put("text", text != null ? text : JSONObject.NULL);
      meta.put("source", "ANDROID_SHARE");
      meta.put("files", files);
      File metaFile = new File(dir, "meta.json");
      try (FileOutputStream out = new FileOutputStream(metaFile)) {
        out.write(meta.toString().getBytes(StandardCharsets.UTF_8));
      }
      // Pending pointer for the WebView / plugin
      File pending = new File(root(), "pending.json");
      JSONObject p = new JSONObject();
      p.put("clientBatchId", batchId);
      p.put("source", "ANDROID_SHARE");
      try (FileOutputStream out = new FileOutputStream(pending)) {
        out.write(p.toString().getBytes(StandardCharsets.UTF_8));
      }
    } catch (Exception e) {
      Log.e(TAG, "Failed to write meta", e);
    }
  }

  public JSONObject loadMeta(String batchId) {
    try {
      File metaFile = new File(batchDir(batchId), "meta.json");
      if (!metaFile.exists()) return null;
      byte[] bytes = readAll(metaFile);
      return new JSONObject(new String(bytes, StandardCharsets.UTF_8));
    } catch (Exception e) {
      return null;
    }
  }

  public JSONObject loadPending() {
    try {
      File pending = new File(root(), "pending.json");
      if (!pending.exists()) return null;
      return new JSONObject(new String(readAll(pending), StandardCharsets.UTF_8));
    } catch (Exception e) {
      return null;
    }
  }

  public void clearPending() {
    new File(root(), "pending.json").delete();
  }

  public void deleteBatch(String batchId) {
    File dir = batchDir(batchId);
    File[] kids = dir.listFiles();
    if (kids != null) {
      for (File f : kids) f.delete();
    }
    dir.delete();
    JSONObject pending = loadPending();
    if (pending != null && batchId.equals(pending.optString("clientBatchId"))) {
      clearPending();
    }
  }

  public List<File> listFiles(String batchId) {
    List<File> out = new ArrayList<>();
    File dir = batchDir(batchId);
    File[] kids = dir.listFiles((d, name) -> !name.equals("meta.json"));
    if (kids != null) {
      java.util.Arrays.sort(kids);
      for (File f : kids) out.add(f);
    }
    return out;
  }

  public String fileToBase64(File f) throws Exception {
    byte[] bytes = readAll(f);
    return Base64.encodeToString(bytes, Base64.NO_WRAP);
  }

  private String queryDisplayName(Uri uri) {
    try (Cursor c = context.getContentResolver().query(
        uri, new String[]{OpenableColumns.DISPLAY_NAME}, null, null, null)) {
      if (c != null && c.moveToFirst()) {
        int idx = c.getColumnIndex(OpenableColumns.DISPLAY_NAME);
        if (idx >= 0) return c.getString(idx);
      }
    } catch (Exception ignored) {
      /* content providers vary */
    }
    return null;
  }

  private static byte[] readAll(File f) throws Exception {
    try (FileInputStream in = new FileInputStream(f)) {
      byte[] buf = new byte[(int) f.length()];
      int off = 0;
      while (off < buf.length) {
        int n = in.read(buf, off, buf.length - off);
        if (n < 0) break;
        off += n;
      }
      return buf;
    }
  }
}
